Extended type assertions

TheRealPiotrP opened the pull request.

Commit:
Adds MethodInfoAssertions.BeAsync() assertion

Commit:
Create MemberInfoAssertions base class to contain assertions that are applicable to all types derived from MemberInfo

Commit:
Moved MethodInfoAssertions onto MemberInfoAssertions base class

Commit:
Move PropertyInfoAssertions onto MemberInfoAssertions

Fix MemberInfoAssertions error string for BeDecoratedWith to take the context of the derived Assertion type into account
Make Context for PropertyInfoAssertions consistent with that for MethodInfoAssertions

Commit:
Added TypeAssertions.Implement(Type interfaceType)

Commit:
Added TypeAssertions.NotHaveProperty(String propertyName)

Commit:
Enables PropertyInfoAssertions.Return

Commit:
Adds PropertyInfoAssertions.BeReadable()

Commit:
Member Accessibility Testing

Create MethodBaseAssertions abstract base class to contain assertions that are applicable to all types derived from MethodBase
Add CSharpAccessModifier and Extensions to enable reasoning about properties' modifiers in C# code
Modify MethodInfoAssertions to derive from MethodBaseAssertions

Commit:
Enables PropertyInfoAssertions.BeReadable(accessModifier, reason, reasonargs)

Commit:
Enabling PropertInfoAssertions.NotBeWritable()

Commit:
PropertyInfoAssertions.BeWritable(accessModifier, because, reasonArgs)

Commit:
PropertyAssertions.NotBeReadable()

Commit:
Enable MethodBase tests on wp and winrt

Commit:
Fix StyleCop errors

Commit:
Adds TypeAssertions.HaveProperty()

Commit:
Adds TypeAssertions.HaveIndexer()

Fixes several issues where only Public members were being identified.

Commit:
Added TypeAssertions.HaveIndexer()

Commit:
Adds TypeAssertions.NotImplement

Adds argument validation to Implement and NotImplement

Commit:
Adds TypeAssertion.NotHaveMethod()

Fixes a bug in parameter type string aggregator.

Commit:
Adds MethodBaseAssertions.Returns and MethodBaseAssertions.ReturnsVoid

Commit:
Adds TypeAssertions.HaveMethod

Commit:
Adds TypeAssertions.HaveConstructor and TypeAssertions.HaveDefaultConstructor

Add ConstructorInfoAssertions to allow asserting about Constructors
Refactor MethodInfo and MethodBase to allow reuse by ConstructorInfoAssertions

Commit:
Add TypeAssertions.BeDerivedFrom

Commit:
Adds TypeAssertion.HaveExplicitMethod

Commit:
Adds TypeAssertion.HaveExplicitMethod

Commit:
Adds TypeAssertion.NotHaveExplicitProperty

Commit:
Adds TypeAssertion.NotHaveExplicitMethod

Commit:
Adds AssemblyAssertions.DefineType

Makes AssemblyAssertions available in all contexts, though limited to DefineType until other methods are tested and adapted.

Commit:
Fix StyleCop Warnings

Commit:
Cleaning up tests

Commit:
Fixes TypeAssertion bugs caused by overloads

Refactors type queries into TypeExtensions
Addresses naming of some TypeExtensions

Commit:
Code cleanup and PropertyInfo.IsVirtual bugfix

Renames CSharpAccessModifiers to CSharpAccessModifier per guidelines
Refactors common code into extension methods
Fixes assumption in PropertyInfo.IsVirtual that the property has a getter
Adds and cleans comments for public assertion methods
Adds parameterTypes to ExplicitMethod assertions

krwq commented on FluentAssertions.Core/Common/CSharpAccessModifier.cs:
```diff
@@ -0,0 +1,12 @@
+    public enum CSharpAccessModifier
+    {
+        Public,
+        Private,
```
test comment

dennisdoomen committed:
Updated the test certificate needed to run the WinRT unit tests.

Fixes #270.

krwq commented on FluentAssertions.Core/Common/TypeExtensions.cs:
```diff
@@ -13,6 +14,9 @@ public static class TypeExtensions
         private const BindingFlags PublicMembersFlag =
             BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance;
```
I bet they didn't mean to add BindingFlags.NonPublic here

TheRealPiotrP replied:
Feel free to open an issue for them :)

krwq commented on FluentAssertions.Core/Common/TypeExtensions.cs:
```diff
@@ -256,14 +260,57 @@ private static bool HasNonPrivateGetter(PropertyInfo propertyInfo)
+
+        public static bool HasExplicitlyImplementedProperty(this Type type, Type interfaceType, string propertyName)
+        {
+            var hasGetter = type.HasParameterlessMethod(string.Format("{0}.get_{1}", interfaceType.FullName, propertyName));
```
I wonder if there is some cleaner way to do it, something like: getter.Parent == type (Parent doesn't exist)

TheRealPiotrP replied:
Explicit implementations are kind of funny. They're not expressed as PropertyInfo but as Getter and Setter methods... it made much of this code a bit ugly.

krwq commented on FluentAssertions.Core/Common/TypeExtensions.cs:
```diff
@@ -256,14 +260,57 @@ private static bool HasNonPrivateGetter(PropertyInfo propertyInfo)
+        public static ConstructorInfo GetConstructor(this Type type, IEnumerable<Type> parameterTypes)
+        {
+            return type
+                .GetConstructors(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static)
```
AllMembersFlag?

TheRealPiotrP replied:
Fixed.

krwq commented on FluentAssertions.Core/Types/MemberInfoAssertions.cs:
```diff
@@ -0,0 +1,84 @@
+            string because = "", params object[] reasonArgs)
+            where TAttribute : Attribute
+        {
+            return BeDecoratedWith<TAttribute>(attr => true, because, reasonArgs);
```
Is it possible to change the other implementation of AndWhichConstraint so that isMatchingAttributePredicate defaults to lambda always returning true? (or perhaps default to null and then inside the method add isMatchingAttributePredicate ?= lambda)

TheRealPiotrP replied:
BeDecoratedWith already existed here before our additions. I don't want to scope-creep to fixing it.

Commit:
Fix cref ambiguity warnings for Type

krwq commented on FluentAssertions.Core/Types/MethodInfoAssertions.cs:
```diff
@@ -86,19 +94,14 @@ public MethodInfoAssertions(MethodInfo methodInfo)
         /// </param>
-        public AndConstraint<MethodInfoAssertions> BeVirtual(
-            string because = "", params object[] reasonArgs)
+        public AndConstraint<MethodBaseAssertions<MethodInfo, MethodInfoAssertions>> Return(Type returnType, string because = "", params object[] reasonArgs)
```
How about `Return<Type>(string because =...)`?

TheRealPiotrP replied:
Done. In fact, I added overloads with Generic Parameters to all assertions that took fixed numbers of type parameters.

krwq commented on FluentAssertions.Core/Types/TypeAssertions.cs:
```diff
@@ -198,6 +200,366 @@ public AndConstraint<TypeAssertions> BeDecoratedWith<TAttribute>(string because
+        /// <param name="because">A formatted phrase as is supported by <see cref="M:System.String.Format(System.String,System.Object[])"/> explaining why the assertion
+        ///             is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.</param>
+        /// <param name="reasonArgs">Zero or more objects to format using the placeholders in <see cref="!:because"/>.</param>
+        public AndConstraint<TypeAssertions> Implement(Type interfaceType, string because = "", params object[] reasonArgs)
```
`Implement<Type>`? (Same question for similar members like NotImplement etc.)

TheRealPiotrP replied:
Done.

krwq commented on FluentAssertions.Core/Types/TypeAssertions.cs:
```diff
@@ -198,6 +200,366 @@ public AndConstraint<TypeAssertions> BeDecoratedWith<TAttribute>(string because
+        /// <param name="reasonArgs">Zero or more objects to format using the placeholders in <see cref="!:because"/>.</param>
+        public AndConstraint<TypeAssertions> Implement(Type interfaceType, string because = "", params object[] reasonArgs)
+        {
+            if (!interfaceType.IsInterface)
```
I think this should also work for abstract classes (same for NotImplements) although this is probably controversial subject. (IMO sounds semantically correct: Type should implement abstract class.)

TheRealPiotrP replied:
The abstract class scenario is handled by BeDerivedFrom. According to MSDN, types derive from types and implement interfaces. There is the special case of a purely abstract base class, but I still like the Derive verb since it implies that we're bound by single inheritance, which is true even in the special case.

krwq commented on FluentAssertions.Core/Types/TypeAssertions.cs:
```diff
@@ -198,6 +200,366 @@ public AndConstraint<TypeAssertions> BeDecoratedWith<TAttribute>(string because
+        /// <param name="because">A formatted phrase as is supported by <see cref="M:System.String.Format(System.String,System.Object[])"/> explaining why the assertion
+        ///             is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.</param>
+        /// <param name="reasonArgs">Zero or more objects to format using the placeholders in <see cref="!:because"/>.</param>
+        public AndConstraint<TypeAssertions> BeDerivedFrom(Type baseType, string because = "", params object[] reasonArgs)
```
I think this should be Type.Should().DeriveFrom (as opposite to Type.Should().BeDerivedFrom), also: `DeriveFrom<Type>`

TheRealPiotrP replied:
I'd like to get some additional opinions here. I chose "Be" because it implies some permanence. DeriveFrom felt a bit more like an ongoing activity. That said, Implements makes sense as-is and doesn't require [cannot handle?] the Be preamble, so maybe trimming it is right.

krwq replied:
As a comparison we don't say BeImplementing :)

krwq commented on FluentAssertions.Core/Types/TypeAssertions.cs:
```diff
@@ -198,6 +200,366 @@ public AndConstraint<TypeAssertions> BeDecoratedWith<TAttribute>(string because
+        /// <param name="reasonArgs">Zero or more objects to format using the placeholders in <see cref="!:because"/>.</param>
+        public AndConstraint<TypeAssertions> BeDerivedFrom(Type baseType, string because = "", params object[] reasonArgs)
+        {
+            if (baseType.IsInterface)
```
For the sake of C++ and fact that you technically derive from the interface I think this should be allowed although it is ok as is too (another controversial subject)

TheRealPiotrP replied:
In terms of C# the language seems unambiguous: https://msdn.microsoft.com/en-us/library/ms228387%28v=vs.90%29.aspx. Classes **derive from** other classes, and inherit their parents' members. Classes **implement** interfaces.

krwq commented on FluentAssertions.Core/Types/TypeAssertions.cs:
```diff
@@ -198,6 +200,366 @@ public AndConstraint<TypeAssertions> BeDecoratedWith<TAttribute>(string because
+        /// <param name="because">A formatted phrase as is supported by <see cref="M:System.String.Format(System.String,System.Object[])"/> explaining why the assertion
+        ///     is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.</param>
+        /// <param name="reasonArgs">Zero or more objects to format using the placeholders in <see cref="!:because"/>.</param>
+        public AndConstraint<TypeAssertions> HaveExplicitProperty(Type interfaceType, string name, string because = "", params object[] reasonArgs)
```
is interfaceType required? It feels like you can do that without interface type. I'd rename current method to ExplicitlyImplementProperty

Commit:
Add missign AssemblyAssertions specs

Commit:
Provide Generics overloads for assertions that take individual Type parameters.

Commit:
Get constructors using AllMembersFlag

Commit:
Fix additional cref ambiguities for System.Type

TheRealPiotrP commented on buildoutput.txt:
```diff
@@ -0,0 +1,239 @@
+psake version 4.3.2
```
get rid of this file...

TheRealPiotrP replied:
Done

TheRealPiotrP commented on FluentAssertions.Core/Common/CSharpAccessModifier.cs:
```diff
@@ -0,0 +1,12 @@
+        Protected,
+        Internal,
+        ProtectedInternal,
+        None
```
Consider a better name... this means that C# cannot interperet the access modifier, but perhaps VB, etc... can

TheRealPiotrP replied:
Done.

TheRealPiotrP commented on FluentAssertions.Core/Common/CSharpAccessModifierExtensions.cs:
```diff
@@ -0,0 +1,68 @@
+            {
+                return CSharpAccessModifier.ProtectedInternal;
+            }
+
```
it (type.IsNotPublic) return CSharpAccessModifier.Internal.

Add test cases for top-level && nested classes.

TheRealPiotrP replied:
Done.

TheRealPiotrP commented on FluentAssertions.Core/Common/MethodInfoExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+
+        internal static bool IsNonVirtual(this MethodInfo method)
+        {
+            return !method.IsVirtual || method.IsFinal;
```
This really means "Is Not Overridable"

TheRealPiotrP commented on FluentAssertions.Core/Common/TypeExtensions.cs:
```diff
@@ -256,14 +260,57 @@ private static bool HasNonPrivateGetter(PropertyInfo propertyInfo)
+            var hasGetter = type.HasParameterlessMethod(string.Format("{0}.get_{1}", interfaceType.FullName, propertyName));
+            var hasSetter = type.GetMethods(AllMembersFlag)
+                .SingleOrDefault(m => m.Name == string.Format("{0}.set_{1}", interfaceType.FullName, propertyName) && m.GetParameters().Count() == 1) != null;
+
```
This is a C# convention. If we have time, let's generalize it...

TheRealPiotrP commented on FluentAssertions.Core/Types/MethodInfoAssertions.cs:
```diff
@@ -32,20 +28,44 @@ public MethodInfoAssertions(MethodInfo methodInfo)
+        /// <param name="because">A formatted phrase as is supported by <see cref="M:System.String.Format(System.String,System.Object[])"/> explaining why the assertion
+        ///             is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.</param>
+        /// <param name="reasonArgs">Zero or more objects to format using the placeholders in <see cref="!:because"/>.</param>
+        public AndConstraint<MethodInfoAssertions> BeAsync(string because = "", params object[] reasonArgs)
```
This is asserting about the implementation, not the signature.

dennisdoomen committed:
Added support for recursively comparing two multi-dimensional arrays for equivalency. Should fix #260.
Fixed a bug when chaining multiple assertions together.
During equivalency assertions, a declared type of object is still processed as a run-time type.

dennisdoomen committed:
Merge pull request #273 from dennisdoomen/MultidimensionalArrays

Added support for recursively comparing two multi-dimensional arrays …

dennisdoomen committed:
Fixed a couple of specs that fail under Silverlight because it can't access private classes.
Disable the Chill tests under Silverlight because disposing doesn't work properly.

dennisdoomen committed:
Added Lighthouse to run the Silverlight tests.

dennisdoomen committed:
Merge pull request #275 from dennisdoomen/SilverlightFixes

Silverlight fixes

Commit:
Addressing MSFT PR feedback
